### PR TITLE
feat: add compatibility mode for push and attach cmd

### PR DESCRIPTION
**Pull request by qweeah**
Resolves #739

### CODE REVIEW

File: cmd/oras/internal/option/spec.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+		opts.ManifestSupportState = registry.OCIArtifact
+	default:
+		return fmt.Errorf("unknown image specification flag: %q", opts.specFlag)
+	}
+	return nil
+}
+
+// ApplyFlags applies flags to a command flag set.
+func (opts *ImageSpec) ApplyFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&opts.specFlag, "image-spec", "", "set OCI image spec version and request manifest type. E.g. `v1.1-image,v1.1-artifact`")
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/spec.go:50]:
```suggestion
	fs.StringVar(&opts.specFlag, "image-spec", "", "specify manifest type for building artifact. options: v1.1-image, v1.1-artifact")
```

Note: The back-tick in the description has special meaning in `cobra`. For example, the code
```go
fs.StringVar(&opts.specFlag, "image-spec", "", "set OCI image spec version and request manifest type. E.g. `v1.1-image,v1.1-artifact`")
```
results in
```
--image-spec v1.1-image,v1.1-artifact    set OCI image spec version and request manifest type. E.g. v1.1-image,v1.1-artifact
```
and not
```
--image-spec string    set OCI image spec version and request manifest type. E.g. `v1.1-image,v1.1-artifact`
```

File: cmd/oras/internal/option/spec.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+	"github.com/spf13/pflag"
+	"oras.land/oras/internal/registry"
+)
+
+// ImageSpec option struct.
+type ImageSpec struct {
+	ManifestSupportState registry.ManifestSupportState
+
+	// should be provided in form of `<version>-<manifest type>`
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/spec.go:29]:
```suggestion
	// specFlag should be provided in form of `<version>-<manifest type>`
```

File: cmd/oras/internal/option/spec.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+// ApplyFlags applies flags to a command flag set.
+func (opts *ImageSpec) ApplyFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&opts.specFlag, "image-spec", "", "set OCI image spec version and request manifest type. E.g. `v1.1-image,v1.1-artifact`")
+}
+
+// DistributionSpec option struct.
+type DistributionSpec struct {
+	ReferrersApiSupportState registry.ReferrersApiSupportState
+
+	// should be provided in form of`<version>-<api>-<option>`
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/spec.go:57]:
```suggestion
	// specFlag should be provided in form of`<version>-<api>-<option>`
```

File: cmd/oras/internal/option/spec.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// ApplyFlags applies flags to a command flag set.
+func (opts *ImageSpec) ApplyFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&opts.specFlag, "image-spec", "", "set OCI image spec version and request manifest type. E.g. `v1.1-image,v1.1-artifact`")
+}
+
+// DistributionSpec option struct.
+type DistributionSpec struct {
+	ReferrersApiSupportState registry.ReferrersApiSupportState
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/spec.go:55]: Having a `ReferrersApiSupportState` seems to heavy and loose for the `DistributionSpec`.

Consider
```suggestion
	// ReferrersAPI indicates the preference of the implementation of the Referrers API. 
	// Set to true for referrers API, false for referrers tag scheme, and nil for auto fallback.
	ReferrersAPI *bool
```

File: cmd/oras/internal/option/spec.go [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		opts.ReferrersApiSupportState = registry.ReferrersApiSupported
+	default:
+		return fmt.Errorf("unknown image specification flag: %q", opts.specFlag)
+	}
+	return nil
+}
+
+// ApplyFlags applies flags to a command flag set.
+func (opts *DistributionSpec) ApplyFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&opts.specFlag, "distribution-spec", "", "set OCI distribution spec version and API option. E.g. `v1.1-referrers-api,v1.1-referrers-tag`")
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/spec.go:78]:
```suggestion
	fs.StringVar(&opts.specFlag, "distribution-spec", "", "set OCI distribution spec version and API option. options: v1.1-referrers-api, v1.1-referrers-tag")
```

File: cmd/oras/push.go [hunk 6]
```diff
@@ -218,7 +230,7 @@ func updateDisplayOption(opts *oras.CopyGraphOptions, store content.Fetcher, ver
 type packFunc func() (ocispec.Descriptor, error)
 type copyFunc func(desc ocispec.Descriptor) error
 
-func pushArtifact(dst *remote.Repository, pack packFunc, packOpts *oras.PackOptions, copy copyFunc, copyOpts *oras.CopyGraphOptions, verbose bool) (ocispec.Descriptor, error) {
+func pushArtifact(dst *remote.Repository, pack packFunc, packOpts *oras.PackOptions, copy copyFunc, copyOpts *oras.CopyGraphOptions, noFallback bool, verbose bool) (ocispec.Descriptor, error) {
```
Comment (shizhMSFT) [inline cmd/oras/push.go:233]: nit: it is more natural to name `allowFallback` instead of `noFallback`.

File: cmd/oras/push.go [hunk 7]
```diff
@@ -243,7 +255,7 @@ func pushArtifact(dst *remote.Repository, pack packFunc, packOpts *oras.PackOpti
 		return root, nil
 	}
 
-	if !copyRootAttempted || root.MediaType != ocispec.MediaTypeArtifactManifest ||
+	if noFallback || !copyRootAttempted || root.MediaType != ocispec.MediaTypeArtifactManifest ||
```
Comment (shizhMSFT) [inline cmd/oras/push.go:258]: nit:
```suggestion
	if !allowFallback || !copyRootAttempted || root.MediaType != ocispec.MediaTypeArtifactManifest ||
```

File: cmd/oras/internal/option/spec.go (hunk 1 above)
Comment (qweeah) [reply cmd/oras/internal/option/spec.go:50]: I though we are using the backtick to show options directly in the flag type, e.g.
```
--image-spec  v1.1-image, v1.1-artifact           specify manifest type for building artifact. options: v1.1-image, v1.1-artifact
```

File: cmd/oras/push.go (hunk 6 above)
Comment (qweeah) [reply cmd/oras/push.go:233]: It's more like `tryFallback`?

File: cmd/oras/internal/option/spec.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+import (
+	"fmt"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+
+	"github.com/spf13/pflag"
+)
+
+const MediaTypeAutoManifest = ""
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/spec.go:26]: Can we remove this line and just use `""`? I think `MediaTypeAutoManifest` creates more confusion than an empty string.

File: cmd/oras/internal/option/spec.go [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+// ImageSpec option struct.
+type ImageSpec struct {
+	// Manifest type for building artifact
+	ManifestMediaType string
+	// specFlag should be provided in form of `<version>-<manifest type>`
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/spec.go:32]: nit: add an empty line between line 31 and 32

File: cmd/oras/internal/option/spec.go [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+// Parse parses flags into the option.
+func (opts *DistributionSpec) Parse() error {
+	switch opts.specFlag {
+	case "":
+		opts.ReferrersAPI = nil
+	case "v1.1-referrers-api", "v1.1-referrers-tag":
+		isApi := opts.specFlag == "v1.1-referrers-api"
+		opts.ReferrersAPI = &isApi
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/spec.go:72]: nit: it would be better to split those two cases.

File: cmd/oras/push.go [hunk 11]
```diff
@@ -73,6 +74,10 @@ Example - Push file "hi.txt" with config type "application/vnd.me.config":
 Example - Push file "hi.txt" with the custom manifest config "config.json" of the custom media type "application/vnd.me.config":
   oras push --config config.json:application/vnd.me.config localhost:5000/hello:latest hi.txt
 
+Example - Push file "hi.txt" and enforce packed manifest type:
+  oras push --image-spec v1.1-image localhost:5000/hello:latest hi.txt # OCI image
+  oras push --image-spec v1.1-artifact localhost:5000/hello:latest hi.txt # OCI artifact
```
Comment (shizhMSFT) [inline cmd/oras/push.go:79]:
```suggestion
  oras push --image-spec v1.1-image localhost:5000/hello:latest hi.txt    # OCI image
  oras push --image-spec v1.1-artifact localhost:5000/hello:latest hi.txt # OCI artifact
```

File: cmd/oras/push.go [hunk 12]
```diff
@@ -99,6 +104,9 @@ Example - Push file "hi.txt" with multiple tags and concurrency level tuned:
 			return option.Parse(&opts)
 		},
 		RunE: func(cmd *cobra.Command, args []string) error {
+			if opts.ManifestMediaType == ocispec.MediaTypeArtifactManifest && opts.manifestConfigRef != "" {
+				return errors.New("cannot pack an OCI artifact with manifest config at the same time")
+			}
```
Comment (shizhMSFT) [inline cmd/oras/push.go:109]: It reminds me that there will be a feature request eventually to support `--config` for `oras attach`.

Comment (qweeah) [reply cmd/oras/push.go:109]: Why? The content of config doesn't matter and the media type of config can be deducted from artifact type.

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: cmd/oras/push.go (hunk 12 above)
Comment (shizhMSFT) [reply cmd/oras/push.go:109]: If they could, they will ask.

File: cmd/oras/internal/option/spec.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+*/
+
+package option
+
+import (
+	"fmt"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+
+	"github.com/spf13/pflag"
```
Comment (shizhMSFT) [inline cmd/oras/internal/option/spec.go:23]: nit:
```suggestion
	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
	"github.com/spf13/pflag"
```

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: cmd/oras/attach.go [hunk 14]
```diff
@@ -51,6 +53,14 @@ func attachCmd() *cobra.Command {
 Example - Attach file 'hi.txt' with type 'doc/example' to manifest 'hello:test' in registry 'localhost:5000'
   oras attach --artifact-type doc/example localhost:5000/hello:test hi.txt
 
+Example - Attach file "hi.txt" with specific manifest type for packing
```
Comment (FeynmanZhou) [inline cmd/oras/attach.go:56]: I would suggest always splitting the sample commands into multiple examples. If you have more than two options under a flag and put all of those sample commands under a flag example, the layout looks too verbose.

File: cmd/oras/push.go [hunk 15]
```diff
@@ -73,6 +74,10 @@ Example - Push file "hi.txt" with config type "application/vnd.me.config":
 Example - Push file "hi.txt" with the custom manifest config "config.json" of the custom media type "application/vnd.me.config":
   oras push --config config.json:application/vnd.me.config localhost:5000/hello:latest hi.txt
 
+Example - Push file "hi.txt" and enforce packed manifest type:
```
Comment (FeynmanZhou) [inline cmd/oras/push.go:77]: Is it more straightforward and clear as follows?
```suggestion
Example - Push file "hi.txt" and enforce building a specific manifest type:
```

File: cmd/oras/push.go [hunk 16]
```diff
@@ -99,6 +104,9 @@ Example - Push file "hi.txt" with multiple tags and concurrency level tuned:
 			return option.Parse(&opts)
 		},
 		RunE: func(cmd *cobra.Command, args []string) error {
+			if opts.ManifestMediaType == ocispec.MediaTypeArtifactManifest && opts.manifestConfigRef != "" {
+				return errors.New("cannot pack an OCI artifact with manifest config at the same time")
```
Comment (FeynmanZhou) [inline cmd/oras/push.go:108]: If this is the stout message for CLI users, I would suggest avoiding using `pack` and using `build a manifest` instead

Comment (qweeah) [reply cmd/oras/push.go:108]: Good catch.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by FeynmanZhou — APPROVED** [review]
LGTM

**Pull request merged**